src/mpid/ch4/generic/am/mpidig_am_recv.h

```diff
@@ -240,61 +250,22 @@ MPL_STATIC_INLINE_PREFIX int MPIDIG_do_irecv(void *buf, MPI_Aint count, MPI_Data
             MPIR_ERR_CHECK(mpi_errno);
             goto fn_exit;
         } else {
-            if (MPIDIG_REQUEST(unexp_req, recv_ready)) {
-                /* if the unexpected recv is ready, the data is in the unexpected buffer. Just
-                 * copy them to complete */
-                mpi_errno = MPIDIG_handle_unexpected(buf, count, datatype, unexp_req);
-                MPIR_ERR_CHECK(mpi_errno);
-                MPID_Request_complete(unexp_req);
-                if (*request == NULL) {
-                    /* Regular (non-enqueuing) path: MPIDIG is responsbile for allocating
-                     * a request. Here we simply return `unexp_req`, which is already completed. */
-                    *request = unexp_req;
-                } else {
-                    /* Enqueuing path: CH4 already allocated request as `*request`.
-                     * Since the real operations has completed in `unexp_req`, here we
-                     * simply copy the status to `*request` and complete it. */
-                    (*request)->status = unexp_req->status;
-                    MPIR_Request_add_ref(*request);
-                    MPID_Request_complete(*request);
-                    /* Need to free here because we don't return this to user */
-                    MPIDI_CH4_REQUEST_FREE(unexp_req);
-                }
+            mpi_errno = MPIDIG_handle_unexpected(buf, count, datatype, unexp_req);
+            MPIR_ERR_CHECK(mpi_errno);
+            MPID_Request_complete(unexp_req);
```

Comment:
In the rndv case, the request is not complete yet. The transport will initiate data transfer and call `target_cmpl_cb` upon finish.

We may need `MPIDIG_handle_unexpected` to return a flag indicating whether it is completed already.